File: divide-and-conquer/median-of-medians/python/MedianOfMedians.py
from math import ceil

def find_median(arr):
    n = len(arr)
    median = find_median_util(arr, n/2+1, 0, n-1)
    print(median)


def find_median_util(arr, k, low, high):
    m = partition(arr, low, high)

    length = m - low + 1

    if length == k:
        return arr[m]

    if length > k:
        return find_median_util(arr, k, low, m-1)
    else:
        return find_median_util(arr, k-length, m+1, high)


def partition(arr, low, high):
    pivot = get_pivot_val(arr, low, high)

    while low < high:
        while arr[low] < pivot:
            low += 1
        while arr[high] > pivot:
            high -= 1

        if arr[low] == arr[high]:
            low += 1

        # swap

        elif low < high:
            temp = arr[low]
            arr[low] = arr[high]
            arr[high] = temp

    return high


def get_pivot_val(arr, low, high):

    if high - low + 1 <= 9:
        sorted(arr)
        return arr[int(len(arr)/2)]

    medians = [0] * int(ceil(( high - low +1 )/5))

    median_index = 0

    while high >= low:

        temp = [0]* min(5, (high - low + 1))

        for i in range(0, len(temp)):
            if low <= high :
                temp[i] = arr[low]
                low += 1

        sorted(temp)

        medians[median_index] = temp[int(len(temp)/2)]
        median_index += 1

    return get_pivot_val(medians, 0, len(medians) - 1)


arr = [25, 24, 33, 39, 3, 18, 19, 31, 23, 49, 45, 16, 1, 29, 40, 22, 15, 20, 24, 4, 13,  34]
find_median(arr)

File: divide-and-conquer/median-of-medians/java/MedianOfMedians.java
package com.basicsstrong.algorithm.divideandconquer;

import java.util.Arrays;

public class MedianOfMedians {
	
	public static void main(String[] args) {
		
		int[] arr = {25, 24, 33, 39, 3, 18, 19, 31, 23, 49, 45, 16, 1, 29, 40, 22, 15, 20, 24, 4, 13,  34};
		
		findMedian(arr);
		
	}

	private static void findMedian(int[] arr) {
		
		int median = findMedianUtil(arr, (arr.length)/2+1, 0, arr.length-1);
		
		System.out.println("Median = "+ median);

		
	}

	private static int findMedianUtil(int[] arr, int k, int low, int high) {
		
		int m = partition(arr, low, high);
		
		int length = m-low+1;
		
		if(length == k) {
			return arr[m];
		}
		
		if(length > k)
			return findMedianUtil(arr, k, low, m-1);
		else
			return findMedianUtil(arr, k-length, m+1, high);
		
	}

	private static int partition(int[] arr, int low, int high) {
		
		
		int pivot = getPivotValue(arr, low, high);
		
		while(low < high) {
			while(arr[low] < pivot)
				low++;
			while(arr[high] > pivot)
				high--;
			
			
			if(arr[low] == arr[high])
				low++;
			
			//swap
			
			else if(low < high) {
				int temp = arr[low];
				arr[low] = arr[high];
				arr[high] = temp;
			}
			
		}
		
		return high;
		
	}

	private static int getPivotValue(int[] arr, int low, int high) {
		
		if(high-low + 1 <= 9) {
			Arrays.sort(arr);
			return arr[arr.length/2];
		}
		
		int[] temp = null;
		
		int medians[] = new int[(int) Math.ceil((double)(high-low+1)/5)];
		
		int medianIndex = 0;
		
		while(high >= low) {
			temp = new int[Math.min(5, high-low+1)];
			
			for (int i = 0; i < temp.length && low <= high; i++) {
				
				temp[i] = arr[low];
				low++;
				
			}
			
			Arrays.sort(temp);
			
			medians[medianIndex] = temp[temp.length/2];
			medianIndex++;
		}
		
		
		return getPivotValue(medians, 0, medians.length-1);
	}

}
